File: src/claw_mem/decay/controller.py
# Copyright 2026 Peter Cheng
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
DecayController - Forgetting controller for graph edge decay.

Manages the lifelong weight calculation, edge classification, and
cleanup of expired edges in the MultiGraphMemory.
"""

import time
import threading
from typing import Dict, List, Tuple

from claw_mem.graph.multi_graph import MultiGraphMemory, SubGraphType
from claw_mem.decay.functions import (
    calculate_weight,
    DecayConfig,
)


class DecayController:
    """Forgetting controller for edge-level decay.

    Responsibilities:
      - Compute decayed edge weights.
      - Classify edges into strength tiers.
      - Decide which edges should be removed.
      - Execute cleanup on the associated MultiGraphMemory.
    """

    def __init__(self, graph: MultiGraphMemory,
                 config: DecayConfig = None):
        self._graph = graph
        self._config = config or DecayConfig.default()
        self._lock = threading.Lock()
        self._last_decay_time: float = 0.0
        self._decay_count: int = 0

    # ── Weight computation ─────────────────────────────────────────

    def calculate_single_weight(self, initial: float, days: float,
                                category: str) -> float:
        """Calculate a single edge weight after decay."""
        return calculate_weight(initial, days, category)

    def get_decay_weight(self, source: str, target: str,
                         edge_type: str, created_at: float) -> float:
        """Compute decayed weight for a specific edge."""
        current_weight = 1.0
        for g in self._graph._graphs.values():
            w = g.edge_weights.get((source, target))
            if w is not None:
                current_weight = w
                break
        days_elapsed = (time.time() - created_at) / 86400.0
        category = self._edge_type_to_category(edge_type)
        return calculate_weight(current_weight, days_elapsed, category)

    def _edge_type_to_category(self, edge_type: str) -> str:
        mapping = {
            "next": "temporal",
            "derived_from": "causal",
            "synthesized_from": "causal",
            "related_to": "semantic",
            "has_concept": "entity",
        }
        return mapping.get(edge_type, "semantic")

    def compute_all_decays(self) -> Dict[Tuple[str, str], float]:
        """Compute decay weights for all edges.

        Returns:
            {(source, target): new_weight} for edges that decayed.
        """
        updates = {}
        now = time.time()

        with self._lock:
            for sg_type, subgraph in self._graph._graphs.items():
                category = {
                    SubGraphType.TEMPORAL: "temporal",
                    SubGraphType.CAUSAL: "causal",
                    SubGraphType.SEMANTIC: "semantic",
                    SubGraphType.ENTITY: "entity",
                }.get(sg_type, "semantic")

                for (s, t), weight in subgraph.edge_weights.items():
                    node = self._graph.get_node(s)
                    if node is not None and hasattr(node, 'created_at') and node.created_at:
                        created_at = node.created_at
                        if not isinstance(created_at, (int, float)):
                            created_at = created_at.timestamp()
                    else:
                        created_at = now
                    days_elapsed = (now - created_at) / 86400.0
                    new_weight = calculate_weight(
                        weight, days_elapsed, category
                    )
                    if new_weight < weight:
                        updates[(s, t)] = max(0.0, min(1.0, new_weight))

        return updates

    # ── Edge classification ───────────────────────────────────────

    def classify_edges(self) -> Dict[str, List[Tuple[str, str, float]]]:
        """Classify all edges by strength tier.

        Returns:
            {"strong": [...], "medium": [...], "weak": [...], "expired": [...]}
        """
        classified = {"strong": [], "medium": [], "weak": [], "expired": []}

        for sg_type, subgraph in self._graph._graphs.items():
            for (s, t), weight in subgraph.edge_weights.items():
                entry = (s, t, weight)
                if weight > self._config.strong_threshold:
                    classified["strong"].append(entry)
                elif weight > self._config.archive_threshold:
                    classified["medium"].append(entry)
                elif weight > self._config.expire_threshold:
                    classified["weak"].append(entry)
                else:
                    classified["expired"].append(entry)

        return classified

    # ── Cleanup ───────────────────────────────────────────────────

    def should_remove_edge(self, source: str, target: str,
                           weight: float) -> bool:
        """Decide whether an edge should be removed."""
        if weight <= self._config.purge_threshold:
            return True
        if weight <= self._config.expire_threshold:
            if self._config.protect_critical:
                node = self._graph.get_node(source)
                if node and getattr(
                    node, 'metadata', {}
                ).get('critical', False):
                    return False
            return True
        return False

    def cleanup_expired(self) -> List[Tuple[str, str]]:
        """Remove expired edges from the graph.

        Returns:
            List of removed (source, target) pairs.
        """
        removed: List[Tuple[str, str]] = []
        with self._lock:
            for sg_type in list(SubGraphType):
                subgraph = self._graph._graphs[sg_type]
                expired = [
                    (s, t) for (s, t), w in subgraph.edge_weights.items()
                    if self.should_remove_edge(s, t, w)
                ]
                removed.extend(expired)

            if removed:
                for (s, t) in removed:
                    for g in self._graph._graphs.values():
                        if g.has_edge(s, t):
                            del g.edge_weights[(s, t)]
                            g.adjacency[s] = [
                                (n, w) for n, w in g.adjacency.get(s, [])
                                if n != t
                            ]
                            g.reverse_adjacency[t] = [
                                (n, w) for n, w in g.reverse_adjacency.get(t, [])
                                if n != s
                            ]
                for g in self._graph._graphs.values():
                    g.edge_count = len(g.edge_weights)

        self._decay_count += 1
        self._last_decay_time = time.time()
        return removed

    # ── Stats ─────────────────────────────────────────────────────

    def get_stats(self) -> dict:
        classified = self.classify_edges()
        total = sum(len(v) for v in classified.values())
        return {
            "total_edges": total,
            "strong_edges": len(classified["strong"]),
            "medium_edges": len(classified["medium"]),
            "weak_edges": len(classified["weak"]),
            "expired_edges": len(classified["expired"]),
            "decay_count": self._decay_count,
            "last_decay_time": self._last_decay_time,
            "config": {
                "purge_threshold": self._config.purge_threshold,
                "expire_threshold": self._config.expire_threshold,
                "decay_interval_hours": self._config.decay_interval_hours,
            },
        }

File: src/deprecated/decay/controller.ts
// Copyright 2026 Peter Cheng
// Licensed under Apache-2.0

/**
 * DecayController - Forgetting controller for graph edge decay.
 *
 * Manages the lifelong weight calculation, edge classification, and
 * cleanup of expired edges in the MultiGraphMemory.
 */

import { calculateWeight, type DecayConfig, DEFAULT_DECAY_CONFIG } from "./functions.js";

// ── External type stubs ────────────────────────────────────────────────

/**
 * Minimal subgraph interface for the parts we use from MultiGraphMemory.
 */
interface SubGraph {
  edgeWeights: Record<string, number>;
  adjacency: Record<string, Array<{ neighbor: string; weight: number }>>;
  reverseAdjacency: Record<string, Array<{ neighbor: string; weight: number }>>;
  edgeCount: number;
  hasEdge(s: string, t: string): boolean;
}

enum SubGraphType {
  TEMPORAL = "temporal",
  CAUSAL = "causal",
  SEMANTIC = "semantic",
  ENTITY = "entity",
}

interface NodeData {
  created_at?: number | Date;
  metadata?: Record<string, unknown>;
}

interface GraphMemory {
  _graphs: Record<string, SubGraph>;
  getNode(id: string): NodeData | null;
  applyDecay(updates: Record<string, number>): void;
}

// ── Controller ─────────────────────────────────────────────────────────

export class DecayController {
  private _lastDecayTime: number = 0;
  private _decayCount: number = 0;

  constructor(
    private _graph: GraphMemory,
    private _config: DecayConfig = DEFAULT_DECAY_CONFIG,
  ) {}

  // ── Weight computation ───────────────────────────────────────────────

  /** Calculate a single edge weight after decay. */
  calculateSingleWeight(initial: number, days: number, category: string): number {
    return calculateWeight(initial, days, category);
  }

  /** Compute decayed weight for a specific edge. */
  getDecayWeight(source: string, target: string, edgeType: string, createdAt: number): number {
    let currentWeight = 1.0;
    for (const g of Object.values(this._graph._graphs)) {
      const w = g.edgeWeights[`${source}\x00${target}`];
      if (w !== undefined) {
        currentWeight = w;
        break;
      }
    }
    const daysElapsed = (Date.now() / 1000 - createdAt) / 86400.0;
    const category = this._edgeTypeToCategory(edgeType);
    return calculateWeight(currentWeight, daysElapsed, category);
  }

  private _edgeTypeToCategory(edgeType: string): string {
    const mapping: Record<string, string> = {
      next: "temporal",
      derived_from: "causal",
      synthesized_from: "causal",
      related_to: "semantic",
      has_concept: "entity",
    };
    return mapping[edgeType] || "semantic";
  }

  /** Compute decay weights for all edges. Returns updates map keyed by "source\x00target". */
  computeAllDecays(): Record<string, number> {
    const updates: Record<string, number> = {};
    const now = Date.now() / 1000;

    const sgTypeToCategory: Record<string, string> = {
      [SubGraphType.TEMPORAL]: "temporal",
      [SubGraphType.CAUSAL]: "causal",
      [SubGraphType.SEMANTIC]: "semantic",
      [SubGraphType.ENTITY]: "entity",
    };

    for (const [sgType, subgraph] of Object.entries(this._graph._graphs)) {
      const category = sgTypeToCategory[sgType] || "semantic";

      for (const [key, weight] of Object.entries(subgraph.edgeWeights)) {
        const sep = key.indexOf("\x00");
        const s = key.slice(0, sep);
        const t = key.slice(sep + 1);

        const node = this._graph.getNode(s);
        let createdAt: number;
        if (node && node.created_at) {
          if (typeof node.created_at === "number") {
            createdAt = node.created_at;
          } else {
            createdAt = node.created_at.getTime() / 1000;
          }
        } else {
          createdAt = now;
        }
        const daysElapsed = (now - createdAt) / 86400.0;
        const newWeight = calculateWeight(weight, daysElapsed, category);
        if (newWeight < weight) {
          updates[key] = Math.max(0.0, Math.min(1.0, newWeight));
        }
      }
    }

    return updates;
  }

  // ── Edge classification ──────────────────────────────────────────────

  /** Classify all edges by strength tier. */
  classifyEdges(): Record<string, Array<{ source: string; target: string; weight: number }>> {
    const classified: Record<string, Array<{ source: string; target: string; weight: number }>> = {
      strong: [],
      medium: [],
      weak: [],
      expired: [],
    };

    for (const subgraph of Object.values(this._graph._graphs)) {
      for (const [key, weight] of Object.entries(subgraph.edgeWeights)) {
        const sep = key.indexOf("\x00");
        const s = key.slice(0, sep);
        const t = key.slice(sep + 1);
        const entry = { source: s, target: t, weight };

        if (weight > this._config.strongThreshold) {
          classified.strong.push(entry);
        } else if (weight > this._config.archiveThreshold) {
          classified.medium.push(entry);
        } else if (weight > this._config.expireThreshold) {
          classified.weak.push(entry);
        } else {
          classified.expired.push(entry);
        }
      }
    }

    return classified;
  }

  // ── Cleanup ──────────────────────────────────────────────────────────

  /** Decide whether an edge should be removed. */
  shouldRemoveEdge(source: string, target: string, weight: number): boolean {
    if (weight <= this._config.purgeThreshold) return true;
    if (weight <= this._config.expireThreshold) {
      if (this._config.protectCritical) {
        const node = this._graph.getNode(source);
        const meta = node?.metadata as Record<string, unknown> | undefined;
        if (node && meta?.critical) return false;
      }
      return true;
    }
    return false;
  }

  /** Remove expired edges from the graph. Returns list of removed (source, target) pairs. */
  cleanupExpired(): Array<{ source: string; target: string }> {
    const removed: Array<{ source: string; target: string }> = [];

    for (const subgraph of Object.values(this._graph._graphs)) {
      for (const [key, weight] of Object.entries(subgraph.edgeWeights)) {
        const sep = key.indexOf("\x00");
        const s = key.slice(0, sep);
        const t = key.slice(sep + 1);
        if (this.shouldRemoveEdge(s, t, weight)) {
          removed.push({ source: s, target: t });
        }
      }
    }

    if (removed.length > 0) {
      for (const { source: s, target: t } of removed) {
        for (const g of Object.values(this._graph._graphs)) {
          if (g.hasEdge(s, t)) {
            delete g.edgeWeights[`${s}\x00${t}`];
            g.adjacency[s] = (g.adjacency[s] || []).filter((n) => n.neighbor !== t);
            g.reverseAdjacency[t] = (g.reverseAdjacency[t] || []).filter((n) => n.neighbor !== s);
          }
        }
      }
      for (const g of Object.values(this._graph._graphs)) {
        g.edgeCount = Object.keys(g.edgeWeights).length;
      }
    }

    this._decayCount++;
    this._lastDecayTime = Date.now() / 1000;
    return removed;
  }

  // ── Stats ────────────────────────────────────────────────────────────

  getStats(): Record<string, unknown> {
    const classified = this.classifyEdges();
    const total =
      classified.strong.length +
      classified.medium.length +
      classified.weak.length +
      classified.expired.length;

    return {
      totalEdges: total,
      strongEdges: classified.strong.length,
      mediumEdges: classified.medium.length,
      weakEdges: classified.weak.length,
      expiredEdges: classified.expired.length,
      decayCount: this._decayCount,
      lastDecayTime: this._lastDecayTime,
      config: {
        purgeThreshold: this._config.purgeThreshold,
        expireThreshold: this._config.expireThreshold,
        decayIntervalHours: this._config.decayIntervalHours,
      },
    };
  }
}
